keps/sig-network/4410-k8s-network-interface/README.md

```diff
@@ -0,0 +1,113 @@
+- Handle support levels like Gateway API (e.g. "core" and "extended")
+- Handle implementation-specific use cases through extension points
+- Decouple the Pod and Node Network setup
+- Provide garbage collection to ensure no resources created during pod setup such as Linux bridges, ebpf programs, 
+allocated IP addresses are left behind after pod deletion
+- Improve the current IP handling for pods (PodIP) to be handle multiple IP addresses and 
+a field to identify the IP address family (IPV4 vs IPV6)
+- Provide backwards compatibility for the existing CNI approach and migration a path to fully adopt KNI
+- Guarantee the network is setup and in a healthy state before containers are started (ephemeral, init, regular)
+- If feasible, provide API awareness of Pod network namespaces (e.g. interface names)
+- Provide a uniform approach for network setup/teardown for both virtualized (kata) and non-virtualized (runc) 
+runtimes including kubevirt. This could eliminate the high and low level runtimes from the networking path
+- Provide a reference implementation of the KNI network runtime
+- Provide the ability to have all the dependencies packaged in the container image (no more CNI binaries in the host file system)
+..- No more downloading CNI binaries via initContainers/Mounting /etc/cni/net.d or /opt/cni/bin
+- Provide the ability to use native k8s resources for configuration such as a ConfigMap's instead of configuration files in host file system
+- Provide an API to indicate network readiness for the node (no more files on disk)
+- Eliminate the need to exec binaries and replace with gRPC
+- Make troubleshooting easier by having logs accessible via kubectl logs
+- Improve network pod startup time
+- Provide the ability to prevent additional scheduling of pods if IPAM is out of IP addresses without evicting running pods
+
+### Non-Goals
+
+1. Any changes to the kube-scheduler 
+2. Any specific implementation other than the reference implementation. However we should ensure the KNI-API is flexible enough to support
+
+## Proposal
+
+The proposal of this KEP is to design and implement the KNI-API and make necessary changes to the CRI-API and container runtimes. The scope should be kept to a minimum and we should target feature parity. 
+
+### User Stories
+
+We are constantly adding these user stories, please join the community sync to discuss. 
+
+#### Story 1
+
+As a cluster operator, I need the ability to determine my network(s) is ready so that my pods come up with a working network.
+
+#### Story 2
+
+As a cluster operator, I need the ability to determine what networks are available on my node so that upstream components can ensure the pod is scheduled on the appropriate node. 
+
+#### Story 3
+
+As a Kubernetes developer, I need the ability to have extension points for pod network setup, teardown and update so that I can support future Kubernetes networking features with either reducing the changes to core kubernetes or eliminating them
```

Comment:
I have a use-case for consideration here that I can break down into 2 parts:

Part 1: Currently the [Rook](https://github.com/rook/rook) project uses Multus to allow storage pods to attach to host network devices without exposing pods on the host network namespace. The goal is to be able to keep as much network isolation for security as possible while also being able to give the Ceph storage platform access to host-local network speeds. It's not clear to me whether the high-level use case I am describing -- attaching pods to specific host devices -- has a place in KNI. I certainly hope so. (As a note here, Rook uses prefers to use macvlan to get subdevices of a host device rather than exposing the host device itself, but I can also imagine systems that want to just expose the host device itself.)

Part 2: Assuming part 1 fits into KNI's purpose, the next part relates to IP assignments. A big hangup the Rook project has with Multus is that we can't easily get static IP assignments on Multus networks. For the Pod network, it is easy to get a static IP via k8s Service, but we don't have the same ease when doing this for dedicated NICs (\*). I would propose that KNI consider whether it can create API flexibility to allow Service functionality on any network, including host-device-attached networks.

(\*): We are aware that [MultusService](https://github.com/k8snetworkplumbingwg/multus-service) exists, but it isn't at a stable enough place for us to requiring it as an add-on for users. Additionally, it requires Rook to

Reply:
@MikeZappa87 this comment is proobing my point on the user story I added as example here https://github.com/kubernetes/enhancements/pull/4477#issuecomment-1944464788

Reply:
I think we need to rethink this user story.

Reply:
>  For the Pod network, it is easy to get a static IP via k8s Service, but we don't have the same ease when doing this for dedicated NICs (*). I would propose that KNI consider whether it can create API flexibility to allow Service functionality on any network, including host-device-attached networks.

if the pod is connected through a external nic to a external network, is the external network responsibility to assign IPs, are we trying to say that kubernetes should manage these external networks ipam? what stops an admin on that network to assign a static ip?